lib/rmg/ObjectManager.cpp

```diff
@@ -0,0 +1,416 @@
+/*
+ * ObjectManager.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "ObjectManager.h"
+#include "CMapGenerator.h"
+#include "TileInfo.h"
+#include "RmgMap.h"
+#include "RoadPlacer.h"
+#include "RiverPlacer.h"
+#include "WaterAdopter.h"
+#include "../CCreatureHandler.h"
+#include "../mapObjects/CommonConstructors.h"
+#include "../mapObjects/MapObjects.h" //needed to resolve templates for CommonConstructors.h
+#include "../mapping/CMap.h"
+#include "../mapping/CMapEditManager.h"
+#include "Functions.h"
+#include "RmgObject.h"
+
+void ObjectManager::process()
+{
+	zone.fractalize();
+		createRequiredObjects();
+}
+
+void ObjectManager::init()
+{
+	DEPENDENCY(WaterAdopter);
+	POSTFUNCTION(RoadPlacer);
+}
+
+void ObjectManager::addRequiredObject(CGObjectInstance * obj, si32 strength)
+{
+	requiredObjects.push_back(std::make_pair(obj, strength));
+}
+
+void ObjectManager::addCloseObject(CGObjectInstance * obj, si32 strength)
+{
+	closeObjects.push_back(std::make_pair(obj, strength));
+}
+
+void ObjectManager::addNearbyObject(CGObjectInstance * obj, CGObjectInstance * nearbyTarget)
+{
+	nearbyObjects.push_back(std::make_pair(obj, nearbyTarget));
+}
+
+void ObjectManager::updateDistances(const rmg::Object & obj)
+{
+	for (auto tile : zone.areaPossible().getTiles()) //don't need to mark distance for not possible tiles
+	{
+		ui32 d = obj.getArea().distanceSqr(tile); //optimization, only relative distance is interesting
+		map.setNearestObjectDistance(tile, std::min((float)d, map.getNearestObjectDistance(tile)));
+	}
+}
+
+const rmg::Area & ObjectManager::getVisitableArea() const
+{
+	return objectsVisitableArea;
+}
+
+int3 ObjectManager::findPlaceForObject(const rmg::Area & searchArea, rmg::Object & obj, std::function<float(const int3)> weightFunction, bool optimizer) const
+{
+	float bestWeight = 0.f;
+	int3 result(-1, -1, -1);
+	
+	for(const auto & tile : searchArea.getTiles())
+	{
+		obj.setPosition(tile);
+		
+		if(!searchArea.contains(obj.getArea()) || !searchArea.overlap(obj.getAccessibleArea()))
+			continue;
+		
+		float weight = weightFunction(tile);
+		if(weight > bestWeight)
+		{
```

Comment:
Checking weight is cheap, so you could do this first and only check area fit for tiles better than current.

Reply:
only positive values of weight function are considered as good positions

Reply:
I'm not sure if weightFunction is cheaper rather than area checks